lmdeploy/vl/model/utils.py

```diff
@@ -92,17 +92,62 @@ def hack_import_with(src: List[str], dst: str = 'torch'):
         sys.modules.pop(item, None)
 
 
-def _set_function(old_func, new_func):
-    """Replace old function with the new function."""
-    import gc
-    refs = gc.get_referrers(old_func)
-    obj_id = id(old_func)
-    for ref in refs:
-        if isinstance(ref, dict):
-            for x, y in ref.items():
-                if id(y) == obj_id:
-                    ref[x] = new_func
-        elif isinstance(ref, MutableSequence):
-            for i, v in enumerate(ref):
-                if id(v) == obj_id:
-                    ref[i] = new_func
+def _set_func(origin_func_path: str,
+              rewrite_func: Callable,
+              origin_func: Callable = None):
+    """Replace old function with the new function.
+
+    Args:
+        origin_func_path (str): original function path
+        rewrite_func (Callable): function to replace with
+        origin_func (Callable): function to replace
+    """
+    # import module
+    split_path = origin_func_path.split('.')
+    for i in range(len(split_path), 0, -1):
+        try:
+            exec('import {}'.format('.'.join(split_path[:i])))
+            break
+        except Exception:
+            continue
+
+    method_class = False
+    if len(split_path) > 1:
+        module_or_class = eval('.'.join(split_path[:-1]))
+        if isinstance(module_or_class, type):
+            method_class = True
+
+    origin_func = eval(origin_func_path) \
+        if origin_func is None else origin_func
+
+    # replace method
+    if not method_class:
+        import gc
+        refs = gc.get_referrers(origin_func)
+        obj_id = id(origin_func)
+        for ref in refs:
+            if isinstance(ref, dict):
+                for x, y in ref.items():
+                    if id(y) == obj_id:
+                        ref[x] = rewrite_func
+            elif isinstance(ref, MutableSequence):
+                for i, v in enumerate(ref):
+                    if id(v) == obj_id:
+                        ref[i] = rewrite_func
+    exec(f'{origin_func_path} = rewrite_func')
+    return origin_func
+
+
+@contextmanager
+def rewrite_ctx(origin_func_path: List[str], rewrite_func: List[Callable]):
```

Comment:
could we add ut for this func since it is used more often?

Reply:
the current implementation doesn't support rewrite transformers dynamic_module well, may add this ut later.